wallet: fix mixed-input transaction accounting in history RPCs (#34872)
bitcoin/bitcoin: doc/release-notes-34872.md

Wallet
------

- Transactions that spend a mix of wallet-owned and non-wallet inputs are now

>   reported safely by the `gettransaction`, `listtransactions`, and

Reviewer: In "doc: add mixed-input history RPC release note" (eba63bfe77c148a447f44ba0703cb211c7a3213d):
I’m not sure how safety is an issue here. How about "reported correctly", "with improved attribution", "in more detail", or similar?
```suggestion
  broken down in more detail by the `gettransaction`, `listtransactions`, and
```

  `listsinceblock` RPCs. Previously such transactions reported a misleading fee

>   and attributed non-wallet outputs as wallet sends (see issue #14136).

Reviewer: In "doc: add mixed-input history RPC release note" (eba63bfe77c148a447f44ba0703cb211c7a3213d):

```suggestion
  and considered all non-wallet outputs to be wallet sends (see issue #14136).
```

>   When every non-wallet input has a known zero value (for example a

Reviewer: In "doc: add mixed-input history RPC release note" (eba63bfe77c148a447f44ba0703cb211c7a3213d):

The "**known** zero value" keeps throwing me for a loop. Transactions can only spend TXOs that are in our UTXO set when we first see them. Therefore, we should always know every spent TXO’s value when we first validate a transaction. Shouldn’t it therefore also be available when we store a transaction in the wallet? If not, the wallet should learn it at the time a block is processed, and the value should be in the revert data?

>   pay-to-anchor output), the wallet can still attribute the full fee and all
>   sent outputs, so these transactions continue to be reported with the usual

Reviewer: In "doc: add mixed-input history RPC release note" (eba63bfe77c148a447f44ba0703cb211c7a3213d):

I think my issue with "attribute" might be that the verb always needs a target for the attribution that I was missing. Maybe it makes sense when we describe it as the wallet being able to attribute it **to itself**?

```suggestion
  pay-to-anchor output), the wallet can still attribute the full fee and all
  sent outputs to itself, so these transactions continue to be reported with the usual
```

>   per-output `send`/`receive` entries. Otherwise the wallet cannot attribute the
>   foreign outputs or its fee share, so it reports a single unattributed

Reviewer: In "doc: add mixed-input history RPC release note" (eba63bfe77c148a447f44ba0703cb211c7a3213d):

```suggestion
  per-output `send`/`receive` entries. Otherwise the wallet cannot compute its share of the
  foreign outputs and the fee, so it reports a single unattributed
```

>   foreign outputs or its fee share, so it reports a single unattributed
>   aggregate `send` summary carrying the negative total of wallet-owned inputs

Reviewer: In "doc: add mixed-input history RPC release note" (eba63bfe77c148a447f44ba0703cb211c7a3213d):

Maybe:

```suggestion
  foreign outputs or its fee share, so it reports the negative total of the wallet-owned inputs
  as the `send` amount
```

  spent, instead of fabricating per-recipient sends or a fee. Wallet-owned
  outputs are still reported as `receive` entries alongside that send summary,
  so deposit detection based on the `receive` category keeps working, and the
  amounts of a transaction's entries sum to the wallet's net change.

  Affected entries additionally expose the new `involves_mixed_inputs`,
  `wallet_debit`, and `wallet_credit` fields, which let consumers detect the
  conservative case without depending on the `category` value. The `vout` field
  is now optional in the result schema: it is omitted from unattributed
  aggregate mixed-input send summaries, which do not correspond to a single

>   output. The `fee` field is likewise omitted when the wallet cannot determine
>   the fee. (#34872)

Reviewer: In "doc: add mixed-input history RPC release note" (eba63bfe77c148a447f44ba0703cb211c7a3213d):

The fee can always be determined per `Σ(inputs)−Σ(outputs)`, rather we cannot determine our share of the fee.

```suggestion
  output. The `fee` field is likewise omitted when the wallet cannot determine its share of
  the fee. (#34872)
```
